Adds basic structure of Stable diffusion and it's usecases. Part of Unit 12 (#164)
johko/computer-vision-course: chapters/en/Unit 5 - Generative Models/Diffusion models/stable_diffusion.md

For example, if you have an image that's 128 pixels wide and tall, it has four times more pixels than an image that's only 64 pixels wide and tall. Because of how self-attention works, dealing with this larger image doesn't just need four times more memory and computing power, it actually needs sixteen times more (since 4 times 4 equals 16). This makes it challenging to work with very high-resolution images, as they require a lot of resources to process. 
Latent diffusion models address the high computational demands of processing large images by using a Variational Auto-Encoder (VAE) to shrink the images into a more manageable size. The idea is that many images have repetitive or unnecessary information. A VAE, after being trained on a lot of data, can compress an image into a much smaller, condensed form. This smaller version still retains the essential features of the original image.

> - How are we linking text and images since we are using prompts?

Reviewer: ```suggestion
- How are we fusing texts with images since we are using prompts?
```

>     As we know that during inference time, we can feed in the description of an image we'd like to see and some pure noise as a starting point, and the model does its best to 'denoise' the random input into something that matches the caption.

Reviewer: ```suggestion
We know that during inference time, we can feed in the description of an image we'd like to see and some pure noise as a starting point, and the model does its best to 'denoise' the random input into something that matches the caption.
```

> SD leverages a pre-trained transformer model based on something called CLIP. CLIP's text encoder was designed to process image captions into a form that could be used to compare images and text, so it is well suited to the task of creating 

Reviewer: Maybe you can give a link to the CLIP chapter

useful representations from image descriptions. An input prompt is first tokenized (based on a large vocabulary where each word or sub-word is assigned a specific token) and then fed through the CLIP text encoder, producing a 768-dimensional
(in the case of SD 1.X) or 1024-dimensional (SD 2.X) vector for each token. To keep things consistent prompts are always padded/truncated to be 77 tokens long, 
and so the final representation which we use as conditioning is a tensor of shape 77x1024 per prompt.

- How can we add-in good inductive biases? 

>     Since, we are trying to generate something new(for eg. a realistic pokemon), we need a way to go beyond the images we have seen before(for eg. a anime pokemon). That's where U-net and self-attention comes into picture. 

Reviewer: ```suggestion
Since we are trying to generate something new(e.g., a realistic Pokemon), we need a way to go beyond the images we have seen before(e.g., an anime pokemon). That's where U-Net and self-attention come into the picture. 
```

Given a noisy version of an image, the model is tasked with predicting the denoised version based on additional clues such as a text description of the image. 

> Ok, so how do we actually feed this conditioning information into the UNet for it to use as it makes predictions? The answer is something called cross-attention. Scattered throughout the UNet are cross-attention layers. 

Reviewer: ```suggestion
Ok, how do we feed this conditioning information into the U-Net to use as it makes predictions? The answer is something called cross-attention. Scattered throughout the U-Net are cross-attention layers. 
```

> Each spatial location in the UNet can 'attend' to different tokens in the text conditioning, bringing in relevant information from the prompt. 

Reviewer: ```suggestion
Each spatial location in the U-Net can 'attend' to different tokens in the text conditioning, bringing in relevant information from the prompt. 
```

> ## How to use txt2img, img2img, in-painting using Diffusers

Reviewer: ```suggestion
## How to Use `text-to-image`, `image-to-image`, Inpainting in Diffusers
```

This section introduces useful usecases and how can we use perform these tasks using Diffusers library.

> - Text to Image(txt2img)

Reviewer: ```suggestion
- `text-to-image`
```

> The idea is to pass in the text prompt and it convert to the image. 

Reviewer: ```suggestion
The idea is to pass in the text prompt, which is converted to the output image. 
```

<iframe
	src="https://hysts-controlnet-v1-1.hf.space/"
	frameborder="0"
	width="850"
	height="450">
</iframe>

> 

Reviewer: Maybe you could add installation of diffusers somewhere below

> Using diffusers library you can get txt2img working in simple 2 steps. For eg. 

Reviewer: ```suggestion
Using the `diffusers` library, you can get `text-to-image` working in 2 steps.
```

> `

Reviewer: ```suggestion
```python
```

from diffusers import AutoPipelineForText2Image
import torch

pipeline = AutoPipelineForText2Image.from_pretrained(
	"runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, variant="fp16"
).to("cuda")
generator = torch.Generator("cuda").manual_seed(31)
image = pipeline("Astronaut in a jungle, cold color palette, muted colors, detailed, 8k", generator=generator).images[0]

> image `

Reviewer: ```suggestion
image
```
```
